File: android/src/main/java/com/lightningkite/khrysalis/observables/binding/PageIndicatorView.binding.actual.kt
@file:Suppress("NAME_SHADOWING")

package com.lightningkite.khrysalis.observables.binding

import com.lightningkite.khrysalis.JsName
import com.lightningkite.khrysalis.observables.*
import com.lightningkite.khrysalis.rx.removed
import com.lightningkite.khrysalis.rx.until
import com.rd.PageIndicatorView


/**
 *
 * Binds the page Indicator to the observable provided and the count.
 * Count will describe how many screens exist, and the selected describes which one
 * will be visible.
 *
 *  Example
 *  val page = StandardObservableProperites(1)
 *  pageView.bind(numberOfPages, page)
 */

fun PageIndicatorView.bind(count: Int = 0, selected: MutableObservableProperty<Int>){
    this.count = count
    selected.subscribeBy{ value ->
        this.selection = value
    }.until(this.removed)
}

/**
 *
 * Binds the page Indicator to the observable provided and the count.
 * Count will describe how many screens exist, and the selected describes which one
 * will be visible.
 *
 *  Example
 *  val page = StandardObservableProperites(1)
 *  pageView.bind(numberOfPages, page)
 */
@JsName("comRdPageIndicatorViewBindDynamic")
fun PageIndicatorView.bind(count: ObservableProperty<Int> , selected: MutableObservableProperty<Int>){

    count.combine(selected){count, selected ->
        return@combine Pair(count,selected)
    }.subscribeBy { (count, selected) ->
        this.count = count
        this.selection = selected
    }.until(this.removed)
}

File: ios/Khrysalis/src/observables/binding/PageIndicatorView.binding.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation
import UIKit


//--- PageIndicatorView.bind(Int, MutableObservableProperty<Int>)
public extension UIPageControl {
    func bind(_ count: Int, _ selected: MutableObservableProperty<Int>) -> Void {
        self.numberOfPages = Int(count)
        var suppress = false
        selected.subscribeBy { value in
            guard !suppress else { return }
            suppress = true
            self.currentPage = Int(value)
            suppress = false
        }.until(self.removed)
        self.addAction(for: .valueChanged, action: {
            guard !suppress else { return }
            suppress = true
            selected.value = Int(self.currentPage)
            suppress = false
        })
    }
    func bind(count: Int, selected: MutableObservableProperty<Int>) -> Void {
        return bind(count, selected)
    }
}

//--- PageIndicatorView.bind(ObservableProperty<Int>, MutableObservableProperty<Int>)
public extension UIPageControl {
    func bind(_ count: ObservableProperty<Int>, _ selected: MutableObservableProperty<Int>) -> Void {
        count.subscribeBy { count in
            self.numberOfPages = Int(count)
        }
        var suppress = false
        selected.subscribeBy { value in
            guard !suppress else { return }
            suppress = true
            self.currentPage = Int(value)
            suppress = false
        }.until(self.removed)
        self.addAction(for: .valueChanged, action: {
            guard !suppress else { return }
            suppress = true
            selected.value = Int(self.currentPage)
            suppress = false
        })
    }
    func bind(count: ObservableProperty<Int>, selected: MutableObservableProperty<Int>) -> Void {
        return bind(count, selected)
    }
}
